File: Cpp/0154_find_minimum_in_rotated_sorted_array_ii.cpp
// https://leetcode.cn/problems/find-minimum-in-rotated-sorted-array-ii/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0, right = nums.size()-1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[right]) left = mid + 1;
            else if (nums[mid] < nums[right]) right = mid;
            else right -= 1;
        }
        return nums[left];
    }
};

int main() {
    Solution sol;
    vector<int> nums = {3,3,3,1};
    cout << sol.findMin(nums) << '\n';
    return 0;
}

File: Python3/0154_find_minimum_in_rotated_sorted_array_ii.py
# https://leetcode.cn/problems/find-minimum-in-rotated-sorted-array-ii/

class Solution:
    def findMin(self, nums: list[int]) -> int:
        left, right = 0, len(nums)-1
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            elif nums[mid] < nums[right]:
                right = mid
            else:
                right -= 1
        return nums[left]